interfaces/system_key.go

```diff
@@ -251,8 +252,45 @@ func SystemKeyMismatch() (bool, error) {
 	diskSystemKey.AppArmorParserFeatures = nil
 	mySystemKey.AppArmorParserFeatures = nil
 
+	ok, err := MatchSystemKeys(mySystemKey, &diskSystemKey)
+	return !ok, err
+}
+
+// RecordedSystemKey returns opaque system key read from the disk.
+func RecordedSystemKey() (interface{}, error) {
+	raw, err := ioutil.ReadFile(dirs.SnapSystemKeyFile)
```

Comment:
(nitpick) we could share the read logic between RecordedSystemKey and SystemKeyMismatch with something like:
```
diff --git a/interfaces/system_key.go b/interfaces/system_key.go
index 2988db52f3..2b9989ccf2 100644
--- a/interfaces/system_key.go
+++ b/interfaces/system_key.go
@@ -214,17 +214,10 @@ func SystemKeyMismatch() (bool, error) {
                return false, err
        }
 
-       raw, err := ioutil.ReadFile(dirs.SnapSystemKeyFile)
-       if err != nil && os.IsNotExist(err) {
-               return false, ErrSystemKeyMissing
-       }
+       diskSystemKey, err := readSystemKey()
        if err != nil {
                return false, err
        }
-       var diskSystemKey systemKey
-       if err := json.Unmarshal(raw, &diskSystemKey); err != nil {
-               return false, err
-       }
        // deal with the race that "snap run" may start, then snapd
        // is upgraded and generates a new system-key with different
        // inputs than the "snap run" in memory. In this case we
@@ -252,28 +245,34 @@ func SystemKeyMismatch() (bool, error) {
        diskSystemKey.AppArmorParserFeatures = nil
        mySystemKey.AppArmorParserFeatures = nil
 
-       ok, err := MatchSystemKeys(mySystemKey, &diskSystemKey)
+       ok, err := MatchSystemKeys(mySystemKey, diskSystemKey)
        return !ok, err
 }
 
-// RecordedSystemKey returns opaque system key read from the disk.
-func RecordedSystemKey() (interface{}, error) {
+func readSystemKey() (*systemKey, error) {
        raw, err := ioutil.ReadFile(dirs.SnapSystemKeyFile)
+       if err != nil && os.IsNotExist(err) {
+               return nil, ErrSystemKeyMissing
+       }
        if err != nil {
-               if os.IsNotExist(err) {
-                       return nil, ErrSystemKeyMissing
-               }
                return nil, err
        }
-
        var diskSystemKey systemKey
        if err := json.Unmarshal(raw, &diskSystemKey); err != nil {
                return nil, err
        }
-
        return &diskSystemKey, nil
 }
 
+// RecordedSystemKey returns opaque system key read from the disk.
+func RecordedSystemKey() (interface{}, error) {
+       diskSystemKey, err := readSystemKey()
+       if err != nil {
+               return nil, err
+       }
+       return diskSystemKey, nil
+}
+
 // CurrentSystemKey calculates and returns opaque system key.
 func CurrentSystemKey() (interface{}, error) {
        currentSystemKey, err := generateSystemKey()
```